vcr/pe/matcher.go

```
@@ -0,0 +1,288 @@
+// If the VC is of the required format, the alg or proofType must also match.
+// vp formats are ignored.
+// This might not be fully interoperable, but the spec at https://identity.foundation/presentation-exchange/#presentation-definition is not clear on this.
+func matchFormat(format *PresentationDefinitionClaimFormatDesignations, credential vc.VerifiableCredential) bool {
+	if format == nil {
+		return true
+	}
+
+	asMap := map[string]map[string][]string(*format)
+	// we're only interested in the jwt_vc and ldp_vc formats
+	if asMap["jwt_vc"] == nil && asMap["ldp_vc"] == nil {
+		return true
+	}
+
+	// only ldp_vc supported for now
+	if entry := asMap["ldp_vc"]; entry != nil {
+		if proofTypes := entry["proof_type"]; proofTypes != nil {
+			for _, proofType := range proofTypes {
+				if matchProofType(proofType, credential) {
+					return true
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+func matchProofType(proofType string, credential vc.VerifiableCredential) bool {
+	proofs, _ := credential.Proofs()
+	for _, p := range proofs {
+		if string(p.Type) == proofType {
+			return true
+		}
+	}
+	return false
+}
+
+func matchDescriptor(descriptor InputDescriptor, credential vc.VerifiableCredential) (*InputDescriptorMappingObject, error) {
+	match, err := matchCredential(descriptor, credential)
+	if err != nil {
+		return nil, err
+	}
+	if !match {
+		return nil, nil
+	}
+
+	return &InputDescriptorMappingObject{
+		Id:     descriptor.Id,
+		Format: "ldp_vc", // todo: hardcoded for now, must be derived from the VC, but we don't support other VC types yet
+	}, nil
+}
+
+func matchCredential(descriptor InputDescriptor, credential vc.VerifiableCredential) (bool, error) {
+	// for each constraint in descriptor.constraints:
+	//   a vc must match the constraint
+	if descriptor.Constraints != nil {
+		return matchConstraint(descriptor.Constraints, credential)
+	}
+	return true, nil
+}
+
+// matchConstraint matches the constraint against the VC.
```

should we error out (or non-match and log a warning) when a VP request uses an unsupported contstraint?

---

that will be up to the caller of the function. This package just implements (some) features of the PresentationExchange spec. I guess the calling piece of code will log a warning/error and return false/403 when it happens.

---

So the matching is a suggestion, but the caller still has to decide whether to actually present the credentials? And if the requester specifies an unsupported constraint, a VC might be matched that doesn't match in reality.

Although this is a larger question: I can imagine this is not a big problem in a mobile wallet, since a person still has to approve the presentation, but on what grounds would a server wallet (without a human making decisions) decide when to/not to present the resulting VCs?

---

This functionality will not return a non-match, it could return an empty list. Then the caller must log and return an error.